### DISCUSSION TITLE
Consider parametrizing RVs in terms of shape

### DISCUSSION THREAD

**Discussion by ricardoV94** [pymc-devs/pytensor#52] [General]
First off, implementation is pretty straightforward, it just requires adding an assert about the support shape of multivariate RVs

These two would be roughly equivalent (but done by the `perform` method)
```python
x = pt.random.dirichlet(np.ones((3, 2)), shape=(10, 3, 2))

x = pt.random.dirichlet(np.ones((3, 2)), size=(10, 3))
# The first two dimensions are already enforced explicitly or follow 
# from broadcasting rules of actions done in the `perform` method
x = at.specify_shape(x, shape=(None, None, 2))  
```

## Cons
* Shape cannot broadcast parameters in the way that size can, because we generally don't know what to do with parameters that determine the support shape. We would simply assert the shape is correct for those dimensions
* It deviates from NumPy semantics

## Pros
* Numpy semantics aren't great. Size means something very different in this context `(shape - support shape)`, whereas usually it's `prod(shape)`. 
  * Note: Numpy does not support broadcasting of multivariate parameters with size.
* It integrates better with PyMC and dims
* Timeseries RVs are more intuitively parametrized with shape, where the last dimension corresponds to number of steps (plus init). As per the current rules of `size`, we always need to break away the last shape dimension into a parameter.
* It "affects" a very small subset of RVs: non-timeseries multivariate which are... Multinomial, MvNormal, MvStudentT, Dirichlet

Related discussions:
* https://github.com/pymc-devs/pymc/discussions/5790
* https://github.com/pymc-devs/pymc/discussions/6090
* https://www.pymc.io/projects/docs/en/stable/learn/core_notebooks/dimensionality.html

**Comment by ricardoV94** [comment]
After #388 this is even more relevant. It is not possible to guess automatically the support shape of a multivariate random variable by definition. Until #388 we made a guess that the last dimensions of the shape of the first argument corresponds to the support shape, which is not a general correct answer and caused confusion when users tried to implement their own RVs, so we removed it. 

This breaks PyMC multivariate `CustomDist`s (or more precisely, breaks the tests that worked because we were creating a dummy MvNormal, whose shape could be inferred with the strategy up to #388). 

From now on, we would need to ask users to give us a callable that returns the support shape. This is really silly because most users will use `shape` or `dims` to define their distributions and that includes the support shape already!. We are artificially discarding this information in some place to ask it again later (just like with timeseries as mentioned above). 

From interacting with RVs for a while I haven't seen any practical use of size. Manipulations are usually done on shapes, and the only time size plays a role is when we need to juggle the shape back into the size argument to recreate an RV.

**Reply by twiecki** [reply]
I think that makes sense. Would also allow us to get rid of a lot of complex code in PyMC.
